### ISSUE TITLE
implementation of seq2seq decoder

### ISSUE THREAD

**Issue by HZhan12** [facebookresearch/wav2letter#407]
Hi, I am working on pytorch implementation of seq2seq model with TDS convs. I've got few questions regarding the decoder and will be very grateful if anyone can clear my confusions.
In the original paper, the decoder is described as follows:

![image](https://user-images.githubusercontent.com/35982842/65506564-6a6b7f00-defe-11e9-9efe-cb6502524cf6.png)

I wonder how I can get K and V from encoder of TDS blocks? In the experiement section, the output dimension of TDS convs is 1024, while the dimension of GRU is 512, so is the output vector(1024D) from encoder split into two 512D vectors(K and V)?

Regarding formula  (4) in the screenshots, how to implement h(S,Q)?

**Comment by an918tw** [comment]
> is the output vector(1024D) from encoder split into two 512D vectors(K and V)?
yes

> Regarding formula (4) in the screenshots, how to implement h(S,Q)?
We add S and Q, apply a linear layer on top and then softmax. (https://github.com/facebookresearch/wav2letter/blob/master/src/criterion/Seq2SeqCriterion.cpp#L272-L275)

**Comment by HZhan12** [comment]
Hi, @an918tw  Thank you for your reply.  Another question regarding  Section 2.2 Efficient Decoder from the paper:
> In principle, this allows us to compute all output frames simultaneously.

Does it mean that with the teacher forcing strategy, Q, S, Prob in Formula(2)(3)(4) can be calculated one by one instead of calculating Qi, Si, Pi (i ∈ 0,1 ...T) recursively？

**Comment by an918tw** [comment]
Yes, with teacher forcing, we know the full text sequence during training, and thus Q in eq. 2 can be compute in one CuDNN call. Eq. 3 will be a matrix multiplication, and with the full S and Q you can compute the probability.

**Comment by HZhan12** [comment]
Hi @an918tw, thank you again for your help. I've got a question regarding hard attention limit in the section of stabilizing beam search.
![image](https://user-images.githubusercontent.com/35982842/67195857-5008c080-f42c-11e9-93c4-7f3748534a5c.png)
I wonder how you implement this?
Given one hypothesis H predicting (y_1, y_2, ..., y_t), assume we have the attention S_t and its peak P_t at time frame t. At time frame t+1, attention S_t+1, computed based on y_t, has the peak P_t+1. If the |P_t+1-P_t|> t_max, then this hypothesis (y_1, y_2,.....y_t) will be removed from the solution candidate set, am I right?

**Comment by an918tw** [comment]
Here's where we have the logic: https://github.com/facebookresearch/wav2letter/blob/master/src/criterion/Seq2SeqCriterion.cpp#L610-L614
Only the valid hypotheses will be added to the candidate set.
